src/API.jl

```diff
@@ -1273,30 +1273,41 @@ function precompile(ctx::Context, pkgs::Vector{PackageSpec}; internal_call::Bool
         end
     end
 
-    stderr_outputs = Dict{Base.PkgId,String}()
+    std_outputs = Dict{Base.PkgId,String}()
     taskwaiting = Set{Base.PkgId}()
     pkgspidlocked = Dict{Base.PkgId,String}()
 
-    function monitor_stderr(pkg, iob)
-        try
-            while isopen(iob)
-                str = readline(iob)
-                stderr_outputs[pkg] = get(stderr_outputs, pkg, "") * str * "\n"
-                if !in(pkg, taskwaiting) && occursin("waiting for IO to finish", str)
-                    !fancyprint && lock(print_lock) do
-                        println(io, pkg.name, color_string(" Waiting for background task / IO / timer.", Base.warn_color()))
+    function monitor_stderr_stdout(pkg, _stderr, _stdout)
+        @sync begin
+            @async try
+                while isopen(_stderr)
+                    str = readline(_stderr)
+                    std_outputs[pkg] = string(get(std_outputs, pkg, ""), str, "\n")
+                    if !in(pkg, taskwaiting) && occursin("waiting for IO to finish", str)
+                        !fancyprint && lock(print_lock) do
+                            println(io, pkg.name, color_string(" Waiting for background task / IO / timer.", Base.warn_color()))
+                        end
+                        push!(taskwaiting, pkg)
                     end
-                    push!(taskwaiting, pkg)
-                end
-                if !fancyprint && in(pkg, taskwaiting)
-                    lock(print_lock) do
-                        println(io, str)
+                    if !fancyprint && in(pkg, taskwaiting)
+                        lock(print_lock) do
+                            println(io, str)
+                        end
                     end
                 end
+            catch err
+                err isa InterruptException || rethrow()
+            end
+            @async try
+                while isopen(_stdout)
+                    str = readline(_stdout)
+                    std_outputs[pkg] = string(get(std_outputs, pkg, ""), str, "\n")
```

Comment:
Do you mean why am I using `readline` rather than `read`? I was using `readline` on stderr so that it's easier to detect `"waiting for IO to finish"`. For stdout read can be used

Reply:
readline is preferable, but any particular reason you are separating the handling now?

Reply:
I think I misunderstood when you said this

> It is strongly forbidden for the same object to be used asynchronously in two difference places.

I took it to mean a single buffer couldn't be used for stderr and stdout. But you meant that it can't be read from async in two places?

Reply:
A Pipe is special, since it is cloned by the kernel instead of userspace